src/gui/components.py

```diff
@@ -18,34 +24,178 @@ def add_qrcode(text, y=QR_PADDING, scr=None, style=None, width=None):
+    def average_chunk_size(msg_size, max_chunk_size):
+        number_of_chunks = ceil(msg_size / max_chunk_size)
+        return ceil(msg_size / number_of_chunks)
+
+    @staticmethod
+    def init(animate):
+        QrA.clean()
+        QrA.animate_ptr = animate
+
+    @staticmethod
+    def run():
+        def QrAtimeout(tim):
+            QrA.timeout = True
+        QrA.timer = pyb.Timer(1, freq=3)
+        QrA.timer.callback(QrAtimeout)
+
+    @staticmethod
+    def qr_cb(obj, event):
+        if event == lv.EVENT.RELEASED:
+            if QrA.isQRtoobig:
+                QrA.isQRplaying = True
+            else:
+                QrA.isQRplaying = not QrA.isQRplaying
+
+    @staticmethod
+    def handle():
+        # call in main loop
+        if QrA.abort:
+            if QrA.timer:
+                QrA.timer.deinit()
+                QrA.timeout = False
+            QrA.isQRplaying = False
+            QrA.abort = False
+        elif QrA.timeout == True:
+            QrA.timeout = False
+            QrA.animate_ptr()
+
+    @staticmethod
+    def pending(text = None):
+        if text == None:
+            return QrA.indx != []
+        return text in QrA.payload or text == "".join(QrA.payload)
+
+    @staticmethod
+    def clean():
+        QrA.isQRplaying = False
+        QrA.indx = []
+        QrA.payload = []
+
+    @staticmethod
+    def stop():
+        # call in event handler of the popup close
+        QrA.abort = True
+
+class QRCode(lv.cont):
+    # max number of bytes to be QR encoded
+    MAX_SIZE = const(300)
+
+    def __init__(self, size, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self._text = "Text"
-        self.set_long_mode(lv.label.LONG.BREAK)
-        self.set_align(lv.label.ALIGN.CENTER)
-        self.set_text(self._text)
-    
-    def set_text(self, text="Text"):
-        if self._text != text:
-            print("QR on screen:", text)
-        self._text = text
-        qr = qrcode.encode_to_string(text)
+        self._text = "txt"
+        self.text_prev = "txt_prev"
+
+        # We have a container that parents a label (QR code)
+        self.set_auto_realign(True)
+        self.align_origo(None, lv.ALIGN.CENTER, 0, 0)
+        self.set_layout(lv.LAYOUT.OFF)
+
+        # label which holds the QR code
+        self.lbl = lv.label(self)
+        self.lbl.set_long_mode(lv.label.LONG.BREAK)
+        self.lbl.set_align(lv.label.ALIGN.CENTER)
+        self.size = size
+        self.max_size = None
+
+        # in case the QR is animatable our container also parents a tip label
+        self.cont = lv.cont(self)
+        self.lbl_indx = lv.label(self.cont)
+        self.lbl_indx.set_text('')
+        self.cont.set_fit(lv.FIT.TIGHT)
+        self.set_fit(lv.FIT.TIGHT)
+        self.isQRbig = False
+        self.animation_tip = True
+
+        QrA.init(self.next)
+        self.set_event_cb(QrA.qr_cb)
+
+    def is_qr_big(self, message):
+        QrA.isQRtoobig = len(message) >= 850
```

Comment:
I set the limit at 850 which may be even lower. It needs more testing. The thing is drawing the qr takes alot of RAM, more than 10 times of the qr payload. One thing i noticed also, we can draw a 1KB qrcode in a 300x300 label most of the times but not into a 480x480 label because it requires that much more RAM.

Reply:
I think to solve it we could move lvgl buffer to SDRAM in the display driver.
I opened a corresponding issue: https://github.com/diybitcoinhardware/f469-disco/issues/8

For now let's just live with this limitation.